data/src/main/java/com/linkedin/data/Data.java

```diff
@@ -131,6 +130,107 @@
     TYPE_MAP.put(ByteString.class, (byte) 9);
   }
 
+  /**
+   * Internal tokens, used to identify types of elements in data during decoding
+   */
+  public enum Token
+  {
+    START_OBJECT,
+    END_OBJECT,
+    START_ARRAY,
+    END_ARRAY,
+    STRING,
+    INTEGER,
+    LONG,
+    FLOAT,
+    DOUBLE,
+    BOOL_TRUE,
+    BOOL_FALSE,
+    NULL,
+    NOT_AVAILABLE
+  }
+
+  /**
+   * Data parser interface invoked by stream decoder.
+   *
+   * This interface contains methods that are invoked when parsing a Data object.
+   * Each method represents a different kind of event/read action
+   *
+   * Methods can throw IOException as a checked exception to
+   * indicate parsing error.
+   *
+   * @author amgupta1
+   */
+  public interface DataParser
```

Comment:
Something about this doesn't sit right with me. I think it would be more appropriate for this interface to be its own top-level class `DataParser` outside of this file. The `Data.Token` enum can probably be it's own thing as well e.g. `DataToken` or even tied with the parser `DataParser.Token`, up to you. Packing everything into this `Data` class doesn't seem intuitive to me.

Reply:
moved it to separate file